process/peer/peerTypeProvider.go

```diff
@@ -156,70 +110,46 @@ func (ptp *PeerTypeProvider) updateCache(epoch uint32) {
 
 func (ptp *PeerTypeProvider) createNewCache(
 	epoch uint32,
-	allNodes map[uint32][]*state.ValidatorInfo,
 ) map[string]*peerListAndShard {
 	newCache := make(map[string]*peerListAndShard)
-	for shardId, validatorsPerShard := range allNodes {
-		for _, v := range validatorsPerShard {
-			newCache[string(v.PublicKey)] = &peerListAndShard{
-				pType:  core.PeerType(v.List),
-				pShard: shardId,
-			}
-		}
-	}
 
 	nodesMapEligible, err := ptp.nodesCoordinator.GetAllEligibleValidatorsPublicKeys(epoch)
 	if err != nil {
 		log.Debug("peerTypeProvider - GetAllEligibleValidatorsPublicKeys failed", "epoch", epoch)
 	}
-	aggregatePType(newCache, nodesMapEligible, core.EligibleList)
+	computePeerTypeAndShardId(newCache, nodesMapEligible, core.EligibleList)
 
 	nodesMapWaiting, err := ptp.nodesCoordinator.GetAllWaitingValidatorsPublicKeys(epoch)
 	if err != nil {
 		log.Debug("peerTypeProvider - GetAllWaitingValidatorsPublicKeys failed", "epoch", epoch)
 	}
-	aggregatePType(newCache, nodesMapWaiting, core.WaitingList)
+	computePeerTypeAndShardId(newCache, nodesMapWaiting, core.WaitingList)
```

Comment:
update cache only if requests and if request is over the time between requests.

Reply:
the update cache method is called here just at the beginning of the epoch so that the heartbeatMonitor has the same values everywhere